src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationService.kt

```diff
@@ -119,36 +120,49 @@ class AllocationService(
     payBandRepository.findByIdOrNull(PayBand.pk(payBandCode))
       ?: throw BadDataException("Pay band code $payBandCode does not exist")
 
-  private fun OffenderProgramProfile.updatePayBands(requestedPayBand: PayBand) {
-    val activePayBand = payBands.first { it.endDate == null }
-    // if we have requested a change to the pay band
-    if (activePayBand.payBand.code != requestedPayBand.code) {
-      val payBandEndingToday = payBands.find { it.endDate == LocalDate.now() }
-      // reinstate the pay band that is ending today
-      if (requestedPayBand.code == payBandEndingToday?.payBand?.code) {
-        payBands.remove(activePayBand)
-        payBandEndingToday.endDate = null
+  private fun OffenderProgramProfile.updatePayBands(requestedPayBand: PayBand, request: UpsertAllocationRequest) {
```

Comment:
Rewritten such that the start/end dates on the Allocation are respected wrt the first and last pay band which have start/end dates.